packages/datadog-plugin-aws-sdk/src/services/kinesis.js

```diff
@@ -1,15 +1,60 @@
 'use strict'
 const {
-  CONTEXT_PROPAGATION_KEY
+  CONTEXT_PROPAGATION_KEY,
+  getSizeOrZero
 } = require('../../../dd-trace/src/datastreams/processor')
 const { encodePathwayContext } = require('../../../dd-trace/src/datastreams/pathway')
 const log = require('../../../dd-trace/src/log')
 const BaseAwsSdkPlugin = require('../base')
+const { storage } = require('../../../datadog-core')
 
 class Kinesis extends BaseAwsSdkPlugin {
   static get id () { return 'kinesis' }
   static get peerServicePrecursors () { return ['streamname'] }
 
+  constructor (...args) {
+    super(...args)
+
+    // TODO(bengl) Find a way to create the response span tags without this WeakMap being populated
+    // in the base class
+    this.requestTags = new WeakMap()
+
+    this.addSub('apm:aws:response:start:kinesis', obj => {
+      const { request, response } = obj
+      const store = storage.getStore()
+      const plugin = this
+      const streamName = this.getStreamName(request.params, request.operation)
+      if (streamName) {
+        this.requestTags.streamName = streamName
```

Comment:
The `streamName` is captured and stored since this argument is used in a function that we aren't particularly interested in tracing (`kinesis.getShardIterator`). The return of this operation is used as the only parameter for `kinesis.getRecords (shardIterator)` and we don't have access to the stream name, which we need when setting the DSM checkpoint. See the example below, which is used in the tests:

`
kinesis.getShardIterator({
    ShardId,
    ShardIteratorType: 'AT_SEQUENCE_NUMBER',
    StartingSequenceNumber: SequenceNumber,
    StreamName: streamName
  }, (err, { ShardIterator } = {}) => {
    if (err) return cb(err)

    kinesis.getRecords({
      ShardIterator
    }, cb)
  })
 `

Reply:
That doesn't appear to be using this though. The only place reading the `streamName` field off the `requestTags` object is later in this same function. 🤔

Reply:
1. We retrieve the `streamName` in this `apm:aws:response:start:kinesis` response function when `kinesis.getShardIterator` or `kinesis.listShards` is called
2. The `streamName` is stored with the plugin object under `requestTags`
3. No other tracing is provided for `kinesis.getShardIterator` and the function call returns
4. `kinesis.getRecords` is called with the `shardIterator` argument (this is a required argument, so we should be safe in assuming that at some point in the past, one of the shardIterator functions was called, at which point we saved the `streamName` to the plugin object).
5. We can access the `streamName` by getting the stored stream name from the plugin object from `this.requestTags.streamName` 
6.  We trace the `getRecords`, setting the DSM checkpoint using the correct `streamName`

Reply:
Ah, I see now what you're doing. I have some concerns though:

1. It's very confusing that multiple operations are handled through the same code here. It's quite unclear that this data is expected to come from a past call due to the externalized functions early-returning on different operation names.
2. Storing streamName on the `requestTags` object like that is problematic as this will not work with concurrent calls. This data would need to be stored in the `storage` to scope correctly.